src/Controls/src/Core/DispatcherExtensions.cs

```diff
@@ -1,68 +1,55 @@
+﻿#nullable enable
 using System;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Maui.Dispatching;
 
 namespace Microsoft.Maui.Controls
 {
 	internal static class DispatcherExtensions
 	{
-		static IDispatcherProvider s_current;
-		static IDispatcher s_default;
-
-		public static IDispatcher GetDispatcher(this BindableObject bindableObject)
+		public static IDispatcher GetDispatcher(this BindableObject? bindableObject)
 		{
-			if (s_default != null)
-			{
-				// If we're already using the fallback dispatcher, keep using it
-				return s_default;
-			}
+			// try find the dispatcher in the current hierarchy
+			if (bindableObject is VisualElement visual &&
+				visual.FindMauiContext() is IMauiContext context &&
+				context.Services.GetService<IDispatcher>() is IDispatcher handlerDispatcher)
+				return handlerDispatcher;
+
+			// try looking on the app
+			if (bindableObject is not Application &&
+				Application.Current?.Dispatcher is IDispatcher appDispatcher)
+				return appDispatcher;
+
+			// maybe this thread has a dispatcher
+			if (Dispatcher.GetForCurrentThread() is IDispatcher globalDispatcher)
+				return globalDispatcher;
```

Comment:
I think that `globalDispatcher` should come before `appDispatcher`: A not-yet-attached bindableObject that is being created for a secondary window (UI thread) should not receive the primary dispatcher.
_(This is the second time I write this comment; earlier, it stayed as "pending" and wasn't visible.)_

Reply:
Good call.